llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -6839,6 +6836,64 @@ static bool ReduceSwitchRange(SwitchInst *SI, IRBuilder<> &Builder,
   return true;
 }
 
+static bool isSwitchOfPowersOfTwo(ArrayRef<APInt> Values) {
+  for (auto &Value : Values) {
+    if (!Value.isPowerOf2())
+      return false;
+  }
+
+  return true;
+}
+
+static bool simplifySwitchOfPowersOfTwo(SwitchInst *SI, IRBuilder<> &Builder,
+                                        const DataLayout &DL) {
+
+  auto *CondTy = cast<IntegerType>(SI->getCondition()->getType());
+
+  if (CondTy->getIntegerBitWidth() > 64 ||
+      !DL.fitsInLegalInteger(CondTy->getIntegerBitWidth()))
+    return false;
+
+  // Only bother with this optimization if there are more than 3 switch cases;
+  // SDAG will only bother creating jump tables for 4 or more cases.
+  if (SI->getNumCases() < 4)
+    return false;
+  SmallVector<APInt, 4> Values;
+  for (const auto &Case : SI->cases())
+    Values.push_back(Case.getCaseValue()->getValue());
+
+  // We perform this optimization only for switches with
+  // unreachable default case.
+  // This assumtion will save us from checking if `Condition` is a power of two
+  bool HasDefault =
+      !isa<UnreachableInst>(SI->getDefaultDest()->getFirstNonPHIOrDbg());
+
+  if (HasDefault || !isSwitchOfPowersOfTwo(Values))
+    return false;
+
+  Builder.SetInsertPoint(SI);
+
+  auto *Condition = SI->getCondition();
+  auto &Context = SI->getContext();
+
+  // FIXME maybe we should check if cttz intrinsic is cheap on the target
```

Comment:
You may need to address it in this PR.

Reply:
I haven't found any way to do this bc at this step we don't have any target machine information. @dtcxzyw suggested using TLI class for that (https://github.com/llvm/llvm-project/issues/70756#issuecomment-1786379626), but, as far as I understand, we are not able to construct it before codegen step.

Reply:
Sorry, TLI is unavailable in the middle end. Can you try `TTI.getIntrinsicInstrCost`?

Reply:
Yeah, I've been thinking about it too, but I could't figure out what to compare this cost with. I also found `TTI::TargetCostConstants` enum but I'm not sure if these are aproppriate values to comare cost with.

Reply:
Well, I believe I've figured it out. We have to compare intrinsic cost with `TTI::TargetCostConstants::TCC_Basic`